nautobot/dcim/filters/__init__.py

```diff
@@ -1826,3 +1836,98 @@ class DeviceTypeToSoftwareImageFileFilterSet(BaseFilterSet):
     class Meta:
         model = DeviceTypeToSoftwareImageFile
         fields = "__all__"
+
+
+class ControllerFilterSet(
+    NautobotFilterSet,
+    LocatableModelFilterSetMixin,
+    TenancyModelFilterSetMixin,
+    LocalContextModelFilterSetMixin,
+    StatusModelFilterSetMixin,
+    RoleModelFilterSetMixin,
+):
+    """Filters for Controller model."""
+
+    q = SearchFilter(
+        filter_predicates={
+            "name": "icontains",
+            "description": "icontains",
+        }
+    )
+    platform = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Platform.objects.all(),
+        to_field_name="name",
+        label="Platform (name or ID)",
+    )
+    external_integration = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=ExternalIntegration.objects.all(),
+        to_field_name="name",
+        label="External integration (name or ID)",
+    )
+    deployed_controller_device = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Device.objects.all(),
+        to_field_name="name",
+        label="Deployed controller device (name or ID)",
+    )
+    deployed_controller_group = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=DeviceRedundancyGroup.objects.all(),
+        to_field_name="name",
+        label="Deployed controller group (name or ID)",
+    )
+
+    class Meta:
+        model = Controller
+        fields = "__all__"
+
+
+class ControllerDeviceGroupFilterSet(NautobotFilterSet):
+    """Filters for ControllerDeviceGroup model."""
+
+    q = SearchFilter(
+        filter_predicates={
+            "name": "icontains",
+        }
+    )
+    controller = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Controller.objects.all(),
+        to_field_name="name",
+        label="Controller (name or ID)",
+    )
+    parent = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=ControllerDeviceGroup.objects.all(),
+        to_field_name="name",
+        label="Parent group (name or ID)",
+    )
+    subtree = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Controller.objects.all(),
+        to_field_name="name",
+        label="Controlled device groups and descendants thereof (name or ID)",
+        method="_subtree",
+    )
+
+    class Meta:
+        model = ControllerDeviceGroup
+        fields = "__all__"
+
+    def generate_query__subtree(self, value):
+        """Helper method used by DynamicGroups and by _subtree() method."""
+        if value:
+            max_depth = (
+                ControllerDeviceGroup.objects.with_tree_fields()
+                .extra(order_by=["-__tree.tree_depth"])
+                .first()
+                .tree_depth
+            )
```

Comment:
Should be able to just use `max_depth = ControllerDeviceGroup.objects.max_depth` property here.

Reply:
Fixed